core/blockchain.go

```diff
@@ -1806,9 +1806,15 @@ func (bc *BlockChain) reprocessState(current *types.Block, reexec uint64) error
 		}
 	}
 
-	for i := 0; i < int(reexec); i++ {
+	for i := 0; i <= int(reexec); i++ {
 		// TODO: handle canceled context
 
+		// Once we have a state, we can reprocess from there.
+		_, err = bc.stateCache.OpenTrie(current.Root())
```

Comment:
It seems we can actually use `HasState` over this and once the loop ends and if there is no parent found with state we can just return a crafted error (like below). I think this would simplify this even further as we don't need to do this extra error handling for no reason.